python/tools/pybridge.py

```diff
@@ -0,0 +1,327 @@
+    return json.dumps(trezor_json)
+
+
+@post("/acquire/<path>/<sid>")  # type: ignore [Untyped function decorator]
+def do_acquire(path: str, sid: str):
+    check_origin()
+    if sid == "null":
+        sid = None  # type: ignore [cannot be assigned to declared type]
+    trezor = Transport.find(path)
+    if trezor is None:
+        response.status = 404
+        return {"error": "invalid path"}
+
+    try:
+        return {"session": trezor.acquire(sid)}
+    except Exception:
+        response.status = 400
+        return {"error": "wrong previous session"}
+
+
+@post("/release/<sid>")  # type: ignore [Untyped function decorator]
+def do_release(sid: str):
+    check_origin()
+    session = Session.find(sid)
+    if session is None:
+        response.status = 404
+        return {"error": "invalid session"}
+    session.release()
+    return {"session": sid}
+
+
+@post("/call/<sid>")  # type: ignore [Untyped function decorator]
+def do_call(sid: str):
+    check_origin()
+    session = Session.find(sid)
+    if session is None:
+        response.status = 404
+        return {"error": "invalid session"}
+
+    msg_type, msg_data = decode_data(request.body.read().decode())
+
+    if msg_type in FILTERS:
+        msg_type, msg_data = FILTERS[msg_type](msg_type, msg_data)
+    log_message("CALLING", msg_type, msg_data)
+
+    session.transport.write(msg_type, msg_data)
+    resp_type, resp_data = session.transport.read()
+
+    if resp_type in FILTERS:
+        resp_type, resp_data = FILTERS[resp_type](resp_type, resp_data)
+    log_message("RESPONSE", resp_type, resp_data)
+
+    return encode_data(resp_type, resp_data)
+
+
+@post("/post/<sid>")  # type: ignore [Untyped function decorator]
+def do_post(sid: str):
+    check_origin()
+    session = Session.find(sid)
+    if session is None:
+        response.status = 404
+        return {"error": "invalid session"}
+
+    msg_type, msg_data = decode_data(request.body.read().decode())
+    session.transport.write(msg_type, msg_data)
+    return {"session": sid}
+
+
+@post("/read/<sid>")  # type: ignore [Untyped function decorator]
+def do_read(sid: str):
+    check_origin()
+    session = Session.find(sid)
+    if session is None:
+        response.status = 404
+        return {"error": "invalid session"}
+
+    resp_type, resp_data = session.transport.read()
+    print("=== RESPONSE:")
+    msg = trezorlib.mapping.DEFAULT_MAPPING.decode(resp_type, resp_data)
+    print(format_message(msg))
+
+    return encode_data(resp_type, resp_data)
+
+
+@post("/listen")  # type: ignore [Untyped function decorator]
+def do_listen():
+    check_origin()
+    try:
+        data = json.load(request.body)
+    except Exception:
+        response.status = 400
+        return {"error": "invalid json"}
+
+    for _ in range(10):
+        trezor_json = [transport.to_json() for transport in Transport.enumerate()]
+        if trezor_json != data:
+            yield json.dumps(trezor_json)
```

Comment:
I'm not sure honestly. It _might_ be a remnat of an earlier version of the code, or it just might be a trick that turns the handler into a generator because we need the listen handler to defer returning the response.
(browser keeps a request to `/listen` open, and bridge is supposed to return data only when the list of devices change (i.e., a trezor is plugged in))

i'll try to replace it with a simple return and see if it still works :woman_shrugging: this code has been sitting on my disk for several months and only now it occured to me to put it in the tools directory where it may be more useful

In any case there's nothing special about it: we yield the data we want to write out, and then end the generator (instead of continuing the for-loop)

Reply:
it's indeed the case: https://bottlepy.org/docs/dev/async.html
added a comment about it in force-push.

i didn't verify whether it _can_ work without the yield, it might, but it's not worth spending time on